torch/_inductor/graph.py

```diff
@@ -943,7 +943,12 @@ def count_bytes(self):
 
     @dynamo_timed
     def compile_to_module(self):
-        from .codecache import PyCodeCache
+        from .codecache import PyCodeCache, warm_pool_once
+
+        # We need to warm process pool lazily.To make sure, even out-of-tree
+        # package's device runtime have been registered before processes are
+        # forked.
+        warm_pool_once()
```

Comment:
I wonder if this will impact compile times... might be worth measuring with an empty triton cache.

Reply:
Let me have a test, I think it is a slight overhead even if it impacts compile times.

Reply:
I make a comparison. there is a slight compilation performance drop, about 1% for model densenet121 in torchvision
for densenet121 cuda model:
```
29.013s  -  warm
42.979s  -  cold, without this patch

29.648s  -  warm
43.379s  -  cold, this patch
```
for densenet121 cpu model:
```
56.833s  -  warm
64.451s  -  cold, without this patch

56.897s  -  warm
64.198s  -  cold, this patch
```
cold compilation is done after running `rm -rf /tmp/torchinductor_$USER`
This is my measure BKC code
```
import torch
import torchvision

# Returns the result of running `fn()` and the time it took for `fn()` to run,
# in seconds. We use CUDA events and synchronization for the most accurate
# measurements.
def timed(fn):
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    result = fn()
    end.record()
    torch.cuda.synchronize()
    return result, start.elapsed_time(end) / 1000

# Generates random input and targets data for the model, where `b` is
# batch size.
def generate_data(b):
    return (
        torch.randn(b, 3, 128, 128).to(torch.float32).cuda(),
        torch.randint(1000, (b,)).cuda(),
    )

N_ITERS = 10

from torchvision.models import densenet121
def init_model():
    return densenet121().to(torch.float32).cuda()

def evaluate(mod, inp):
    with torch.no_grad():
        return mod(inp)

model = init_model()

# Reset since we are using a different mode.
import torch._dynamo
torch._dynamo.reset()

evaluate_opt = torch.compile(evaluate, mode="reduce-overhead")

inp = generate_data(16)[0]
print("eager:", timed(lambda: evaluate(model, inp))[1])
print("compile:", timed(lambda: evaluate_opt(model, inp))[1])
```

Reply:
The DDP ut will fail when we warm the pool lazily. So we changed back the code. Nowadays, each third-party backend uses its own process pool.